remount/remount.c

```diff
@@ -0,0 +1,605 @@
+		strncpy(p->exp + 1, prefix, prefix_len);
+	// add regex key
+	strncpy(p->exp + prefix_len + 1, json + (*k)->start, n);
+	// force end after key regex and remove trailing /
+	if (p->exp[prefix_len + n] == '/')
+		p->exp[prefix_len + n] = '$';
+	else
+		p->exp[prefix_len + n + 1] = '$';
+
+	// parse options value
+	n = (*k + 1)->end - (*k + 1)->start;
+	p->opts = calloc(n + 1, sizeof(char));
+	snprintf(p->opts, n + 1, "%s", json + (*k + 1)->start);
+
+	printf("DEBUG: parsed remount regex '%s' with opts '%s'...\n\r",
+	       p->exp, p->opts);
+
+	ret = 0;
+
+out:
+	jsmnutil_tokv_free(k);
+
+	if (tokv)
+		free(tokv);
+}
+
+void parse_remount_array(char *json, struct dl_list *remount_entries)
+{
+	printf("DEBUG: parsing remount array...\n\r");
+
+	char *str = NULL;
+	int tokc, size;
+	jsmntok_t *tokv, *t;
+
+	if (jsmnutil_parse_json(json, &tokv, &tokc) < 0)
+		goto out;
+
+	size = jsmnutil_array_count(json, tokv);
+	if (size <= 0)
+		goto out;
+
+	t = tokv + 1;
+	while ((str = pv_json_array_get_one_str(json, &size, &t))) {
+		printf("DEBUG: remount entry object '%s' found\n\r", str);
+
+		parse_remount_entry(str, remount_entries);
+		free(str);
+		str = NULL;
+		t = t + 2;
+	}
+
+out:
+	if (str)
+		free(str);
+	if (tokv)
+		free(tokv);
+}
+
+void parse_remount_object(char *json, struct dl_list *remount_entries)
+{
+	printf("DEBUG: parsing remount object...\n\r");
+
+	int tokc;
+	jsmntok_t *tokv = NULL;
+
+	if (jsmnutil_parse_json(json, &tokv, &tokc) < 0) {
+		printf("WARN: JSON could not be parsed\n\r");
+		goto out;
+	}
+
+	char *policy = NULL;
+	policy = getenv("PV_REMOUNT_POLICY");
+	if (!policy)
+		policy = getenv("pv_remount_policy");
+	if (!policy)
+		policy = "default";
+	printf("DEBUG: remount policy '%s' to be used\n\r", policy);
+
+	char *remount_array = NULL;
+	remount_array = pv_json_get_value(json, policy, tokv, tokc);
+	if (!remount_array) {
+		printf("WARN: '%s' key not found\n\r", policy);
+		goto out;
+	}
+
+	parse_remount_array(remount_array, remount_entries);
+	free(remount_array);
+
+out:
+	if (tokv)
+		free(tokv);
+}
+
+void parse_device_run_object(char *json, struct dl_list *remount_entries)
+{
+	printf("DEBUG: parsing component JSON...\n\r");
+
+	int tokc;
+	jsmntok_t *tokv = NULL;
+
+	if (jsmnutil_parse_json(json, &tokv, &tokc) < 0) {
+		printf("WARN: JSON could not be parsed\n\r");
+		goto out;
+	}
+
+	char *remount_object = NULL;
+	remount_object = pv_json_get_value(json, "remount", tokv, tokc);
+	if (!remount_object) {
+		printf("WARN: remount key not found\n\r");
+		goto out;
+	}
+
+	parse_remount_object(remount_object, remount_entries);
+	free(remount_object);
+
+out:
+	if (tokv)
+		free(tokv);
+}
+
+char *get_run_json_rel_path(void)
+{
+	char *pname = NULL;
+	pname = getenv("LXC_NAME");
+	if (!pname) {
+		printf("ERROR: LXC_NAME env not set\n\r");
+		return NULL;
+	}
+	printf("DEBUG: container name '%s' to be used\n\r", pname);
+
+	char *path_fmt = "%s/run.json";
+	size_t len = strlen(path_fmt) + strlen(pname) + 1;
```

Comment:
This is not the right size because you are adding `path_fmt + pname + 1` and the string in `pname` replaces 2 chars in `path_fmt` should be `path_fmt - 2 + pname + 1`, but here is the example where two `snprintf` are needed. Other option is to use a fixed buffer with size `PATH_MAX` and `strdup()`.
I don't know if this could be a source of problems in the future...

Reply:
I think now it's ok